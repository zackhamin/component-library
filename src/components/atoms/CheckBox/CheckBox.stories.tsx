import { CheckBox, ICheckBoxProps } from './CheckBox';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'atoms/CheckBox',
	component: CheckBox,
} as Meta;

const Template: Story<ICheckBoxProps> = (args) => {
	return <CheckBox {...args}>Child</CheckBox>;
};

export const defaultStory = Template.bind({});
defaultStory.args = {
	children: 'Test',
	value: 'Value 1',
	isDisabled: false,
	id: 'id3',
	name: 'name1',
};

export const disabledOption = Template.bind({});
disabledOption.args = {
	children: 'Test Disabled',
	value: 'Value 1',
	isDisabled: true,
	id: 'id4',
	name: 'name2',
};
