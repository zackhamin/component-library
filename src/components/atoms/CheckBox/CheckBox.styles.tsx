import { colors } from 'Styles';
import styled from 'styled-components';
import { typography } from 'Styles';
import { visuallyHiddenStyles } from 'Styles';

const StyledWrapper = styled.div`
	position: relative;
`;

const StyledCheckBoxLabel = styled.label`
	font-family: ${typography.type.body};
	font-weight: 500;
	font-size: ${typography.size.s3 + 'rem'};
	margin-left: 2.8rem;
	margin-right: 1.8rem;
	display: inline-block;
	cursor: pointer;
	margin-top: 3rem;
	&::before {
		content: '';
		position: absolute;
		display: inline-block;
		border-radius: 25%;
		border: 1px solid ${colors.mediumGrey};
		left: -1.75rem;
		top: 50%;
		border-radius: 22%;
		width: 2.1rem;
		height: 2.1rem;
		background: transparent;
		transform: translateY(-50%);
		margin-left: 1.5rem;
		margin-top: 1.5rem;
	}
`;

const StyledCheckIcon = styled.div`
	position: absolute;
	width: 2rem;
	height: 2rem;
	display: none;
	top: 5px;
	left: 0.5px;
	margin-top: 2.7rem;
`;

const StyledCheckBox = styled.input`
	${visuallyHiddenStyles}
	&:checked ~ label {
		& > div {
			display: block;
		}

		&::before {
			content: '';
			border-radius: 22%;
			background: ${colors.blue};
			border: 1px solid ${colors.cobaltBlue};
		}
	}

	&:focus {
		+ ${StyledCheckBoxLabel}::before {
			outline: 1px solid ${colors.neonBlue};
			outline-offset: 2px;
		}
	}

	&:disabled ~ ${StyledCheckBoxLabel} {
		cursor: default;
		&::before {
			background: ${colors.lightGrey};
			cursor: default;
			border: 1px solid ${colors.grey400};
		}
	}
`;

export { StyledWrapper, StyledCheckIcon, StyledCheckBox, StyledCheckBoxLabel };
