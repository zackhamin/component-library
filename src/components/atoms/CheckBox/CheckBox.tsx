import { Check } from 'Atoms/Icons';
import { CheckBoxContext } from 'Molecules/CheckBoxGroup/CheckBoxGroup';
import { ReactInputProps } from '@Types/form';
import React, { forwardRef, useContext } from 'react';
import {
	StyledCheckBox,
	StyledCheckBoxLabel,
	StyledCheckIcon,
	StyledWrapper,
} from './CheckBox.styles';

export interface ICheckBoxProps extends Omit<ReactInputProps, 'ref'> {
	children: React.ReactNode;
	value: string;
	onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
	isDisabled?: boolean;
	isChecked?: boolean;
	name?: string;
	id: string;
}

const CheckBox = forwardRef<HTMLInputElement, ICheckBoxProps>(
	(
		{ children, value, isDisabled, isChecked, onChange, name, id, ...rest },
		ref
	) => {
		const ctx = useContext(CheckBoxContext);
		const checkBoxCtx: any = {
			onChange: ctx?.handleOnChange || onChange,
		};

		return (
			<StyledWrapper>
				<StyledCheckBox
					type="checkbox"
					id={id}
					name={name}
					value={value}
					disabled={isDisabled}
					checked={isChecked}
					onChange={checkBoxCtx.onChange}
					ref={ref}
					{...rest}
				/>

				<StyledCheckBoxLabel htmlFor={id}>
					<StyledCheckIcon>
						<Check color={isDisabled ? 'rgba(0, 0, 0, 0.25)' : 'white'} />
					</StyledCheckIcon>
					{children}
				</StyledCheckBoxLabel>
			</StyledWrapper>
		);
	}
);

export { CheckBox };
