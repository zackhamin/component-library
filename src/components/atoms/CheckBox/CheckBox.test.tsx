import { CheckBox } from './CheckBox';
import { fireEvent, render, screen } from '@testing-library/react';

const defaultProps = {
	value: 'opt1',
	isDisabled: false,
	onChange: jest.fn(),
	children: 'label1',
	id: '1',
};

describe('CheckBox', () => {
	it('renders  correctly', () => {
		const { container } = render(<CheckBox {...defaultProps} />);
		expect(container).toMatchSnapshot();
	});
	it('renders with disabled state', () => {
		render(<CheckBox {...defaultProps} isDisabled={true} />);
		expect(screen.getByRole('checkbox')).toBeDisabled();
	});
});

describe('Checkbox change handlers', () => {
	it('fires change events on change of value', () => {
		render(<CheckBox {...defaultProps} />);
		const input = screen.getByRole('checkbox');
		fireEvent.click(input, { target: { value: 'opt1' } });
		expect({ ...defaultProps.onChange }).toHaveBeenCalled();
	});
});

describe('Checkbox checked state', () => {
	it('should render unchecked', () => {
		render(<CheckBox {...defaultProps}>Checkbox</CheckBox>);

		const checkbox = screen.getByRole('checkbox');
		expect(checkbox).not.toBeChecked();
	});
	it('should be checked on check', () => {
		render(<CheckBox {...defaultProps}>Checkbox</CheckBox>);

		const input = screen.getByRole('checkbox');
		fireEvent.click(input, { target: { value: 'opt1' } });
		const checkbox = screen.getByRole('checkbox');
		expect(checkbox).toBeChecked();
	});
	it('should be checked on render if isChecked is true', () => {
		render(
			<CheckBox {...defaultProps} isChecked={true}>
				Checkbox
			</CheckBox>
		);

		const checkbox = screen.getByRole('checkbox');
		expect(checkbox).toBeChecked();
	});
});
