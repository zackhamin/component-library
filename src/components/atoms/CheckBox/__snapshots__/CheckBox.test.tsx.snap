// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox renders  correctly 1`] = `
.c0 {
  position: relative;
}

.c3 {
  font-family: Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 1.6rem;
  margin-left: 2.8rem;
  margin-right: 1.8rem;
  display: inline-block;
  cursor: pointer;
  margin-top: 3rem;
}

.c3::before {
  content: '';
  position: absolute;
  display: inline-block;
  border-radius: 25%;
  border: 1px solid #bebebe;
  left: -1.75rem;
  top: 50%;
  border-radius: 22%;
  width: 2.1rem;
  height: 2.1rem;
  background: transparent;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  margin-left: 1.5rem;
  margin-top: 1.5rem;
}

.c4 {
  position: absolute;
  width: 2rem;
  height: 2rem;
  display: none;
  top: 5px;
  left: 0.5px;
  margin-top: 2.7rem;
}

.c1 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
  white-space: nowrap;
  word-wrap: normal;
}

.c1:checked ~ label > div {
  display: block;
}

.c1:checked ~ label::before {
  content: '';
  border-radius: 22%;
  background: #0000FF;
  border: 1px solid #0047AB;
}

.c1:focus + .c2::before {
  outline: 1px solid #1F51FF;
  outline-offset: 2px;
}

.c1:disabled ~ .c2 {
  cursor: default;
}

.c1:disabled ~ .c2::before {
  background: #EEEEEF;
  cursor: default;
  border: 1px solid #CDCDCD;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      id="1"
      type="checkbox"
      value="opt1"
    />
    <label
      class="c2 c3"
      for="1"
    >
      <div
        class="c4"
      >
        <svg
          aria-hidden="true"
          color="white"
          focusable="false"
          height="1em"
          viewBox="0 0 18 14"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M17.707.793a1 1 0 0 1 0 1.414l-11 11a1 1 0 0 1-1.414 0l-5-5a1 1 0 0 1 1.414-1.414L6 11.086 16.293.793a1 1 0 0 1 1.414 0Z"
            fill="currentColor"
          />
        </svg>
      </div>
      label1
    </label>
  </div>
</div>
`;
